### PR TITLE
Graph partitioning

### PR DESCRIPTION
**Pull request by bertmaher**
This PR is a placeholder for the graph partitioning work; since it's a big chunk of design (though not a ton of code, yet) I want to be able to get some incremental feedback on the implementation.

So far, this code simply splits a Function into many smaller, possibly parallel, Functions, where each sub-function is strictly sequential.  I haven't done the plumbing to hook up Variables to the input/output, or any code generation yet.

### CODE REVIEW

**Comment by nadavrot** [comment]
The split functions done have to be sequential. They can be any dag.

**Comment by bertmaher** [comment]
Woo, running a partitioned graph works!  Of course, this assumes that the list produced by `partition` is topologically sorted, and that you run the Functions serially.  But, baby steps.

**Comment by nadavrot** [comment]
Cool!

**Comment by jfix71** [comment]
This is awesome!! Looks pretty great to me. Just a couple initial thoughts:

For now this is for partitioning a Function to many of the same kinds of backend. I'm assuming we will want to be able to partition across multiple backend kinds in the future. Perhaps we eventually will pass in a list of available backends to `partition()` then instead of a single one (which is also currently unused anyway). 

Assigning nodes to tasks/backends then will be partially dependent on which backends support which operators. Related, I think we will want to partition before quantizing, with partition querying `isOpSupported()`, so that we do not assign some node we want to quantize to a backend that cannot support it. And in addition to just performance reasons for selecting a specific partitioning, maybe we will want to consider precision loss due to different kinds of quantization support on different backends?

**Comment by stoklund** [comment]
I am a little worried about the design of the `Backend` class hierarchy. It's not so much what you're doing here, but `Backend` is already a bit of a ["The Blob" antipattern](https://sourcemaking.com/antipatterns/the-blob), and this change adds more functionality. This makes it harder to refactor later.

Could we break up `Backend`'s responsibilities before we add more functionality? I'd be happy to help.

The `Backend` subclass objects currently are responsible for:

1. Providing information about the target architecture and settings.
1. Compiling a function for the target.
1. Storing the compiled function.
1. Storing data needed by the function at runtime.
1. Executing the compiled function and copying data to and from the device.
1. (with this change) Scheduling and executing a graph of dependent functions.
1. (internally to some impls) Holding temporary data needed during compilation.

In particular, the different types of state stored by the `Backend` instances is problematic when we have multiple functions and multiple execution units / devices.

**Comment by bertmaher** [comment]
@jfix71 That's a great point.  Part of the reason for the vestigial Backend& param was exactly so that the backend could guide partitioning, but then I ended up not doing anything with it :-p.  I'm kind of unsure what the ExecutionEngine interface should even look like in a multi-backend world (which is why this PR sort of adds a separate concurrent execution path).

**Comment by bertmaher** [comment]
@stoklund I completely agree.  Want to spawn an issue for it and we can coordinate a bit?  I don't have a burning need to land this PR as-is so it'd be nice to do the refactoring to make it nice beforehand.

**Comment by bertmaher** [comment]
OK, I've refactored this to take advantage of the new `Backend` interface.  There are a few interesting pieces here:
- Partitioning is handled in a backend-specific way.  I think this makes sense, because any given backend is probably going to want different partitioning parameters.
- `Backend::partition` creates a `FunctionGraph` that holds the sub-functions and the dependences between them.
- A new `MultiCPUBackend` knows how to partition a graph maximally into parallel tasks...
- And it compiles a partitioned graph into a `MultiCPUFunction` that knows how to execute it.  The current implementation is fairly simple, using condition variables to wait on dependences.
- All of this is completely invisible to the end user, other than creating an `ExecutionEngine(BackendKind::MultiCPU)`.  The EE internally knows whether a backend supports partitioning, and invokes that process if so.

File: lib/Backends/CPU/MultiCPUFunction.cpp [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+  std::vector<SyncVariable> sync(G_.getChannels().size());
+
+  // For each "backend" (representing a partitioned function):
+  //   Wait for its input dependencies to be satisfied.
+  //   Execute it.
+  //   Signal that its output dependencies are satisfied.
+  std::mutex m;
+  auto cfit = functions_.begin();
+  auto end = functions_.end();
+  auto ffit = G_.getFunctions().begin();
```
Comment (qcolombet) [inline lib/Backends/CPU/MultiCPUFunction.cpp]: ffit?
I don't get the second f.
A more descriptive name would be helpful.

File: include/glow/Graph/FunctionGraph.h [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+  /// Input varabiels of each function.
+  std::unordered_map<Function *, VariablesList> inputs;
+
+  /// Output variables of each function.
+  std::unordered_map<Function *, VariablesList> outputs;
+
+public:
+  /// Ctor.
+  FunctionGraph(FunctionList fs, std::unordered_set<Variable *> vs);
```
Comment (qcolombet) [inline include/glow/Graph/FunctionGraph.h]: Could we use more descriptive names for the arguments.

Comment (qcolombet) [inline include/glow/Graph/FunctionGraph.h]: Shouldn't we use references for the arguments.

File: include/glow/Backends/Backend.h [hunk 3]
```diff
@@ -43,6 +45,11 @@ class Backend {
   virtual std::unique_ptr<CompiledFunction>
   compile(std::unique_ptr<IRFunction> IR) const = 0;
 
+  virtual std::unique_ptr<CompiledFunction>
+  compile(std::vector<std::unique_ptr<IRFunction>> IR, FunctionGraph &G) const {
```
[the remark this answers is not in the document]
Comment (bertmaher) [reply include/glow/Backends/Backend.h]: Yeah, since it's a vector-of-unique-ptrs the copy ctor is deleted anyways, but I'll make it `&&` for clarity.

**Comment by bertmaher** [comment]
A good idea from conversation with @opti-mix: make the test use bigger FCs and initialize the weights outside the test proper, so the speedup from multithreading is visible.

**Comment by bertmaher** [comment]
Where should execute() go?  It seems like the code to execute a partitioned graph is necessarily backend-specific.

**Comment by bertmaher** [comment]
Based on feedback I'm going to abandon this approach and defer the implementation until a design review can be conducted.

**Pull request closed without merging**